c/misra/src/rules/RULE-21-25/InvalidMemoryOrderArgument.ql

```diff
@@ -0,0 +1,172 @@
+/**
+ * @id c/misra/invalid-memory-order-argument
+ * @name RULE-21-25: All memory synchronization operations shall be executed in sequentially consistent order
+ * @description Only the memory ordering of 'memory_order_seq_cst' is fully portable and consistent.
+ * @kind path-problem
+ * @precision very-high
+ * @problem.severity error
+ * @tags external/misra/id/rule-21-25
+ *       external/misra/c/2012/amendment4
+ *       correctness
+ *       concurrency
+ *       external/misra/obligation/required
+ */
+
+import cpp
+import codingstandards.c.misra
+import semmle.code.cpp.dataflow.new.DataFlow
+
+/* A member of the set of memory orders defined in the `memory_order` enum */
+class MemoryOrder extends EnumConstant {
+  MemoryOrder() { getDeclaringEnum().getName() = "memory_order" }
+
+  int getIntValue() { result = getValue().toInt() }
+}
+
+/* This is the only standardized memory order, allowed by RULE-21-25. */
+class AllowedMemoryOrder extends MemoryOrder {
+  AllowedMemoryOrder() { getName() = "memory_order_seq_cst" }
+}
+
+/* An expression referring to a memory order */
+class MemoryOrderConstantAccess extends EnumConstantAccess {
+  MemoryOrderConstantAccess() { getTarget() instanceof MemoryOrder }
+
+  predicate isAllowedOrder() { getTarget() instanceof AllowedMemoryOrder }
+}
+
+/* An expression with a constant value that equals a `MemoryOrder` constant */
+class MemoryOrderConstantExpr extends Expr {
+  MemoryOrder ord;
+
+  MemoryOrderConstantExpr() {
+    if this instanceof MemoryOrderConstantAccess
+    then ord = this.(MemoryOrderConstantAccess).getTarget()
+    else ord.getIntValue() = getValue().toInt()
+  }
+
+  /* Get the name of the `MemoryOrder` this expression is valued as. */
+  string getMemoryOrderString() { result = ord.toString() }
+}
+
+/**
+ * A `stdatomic.h` function which accepts a `memory_order` value as a parameter.
+ */
+class MemoryOrderedStdAtomicFunction extends Function {
+  int orderParamIdx;
+
+  MemoryOrderedStdAtomicFunction() {
+    exists(int baseParamIdx, int baseParams, string prefix, string suffix |
+      prefix = ["__", "__c11_"] and
+      suffix = ["", ".*", "_explicit"] and
+      (
+        getName().regexpMatch(prefix + ["atomic_thread_fence", "atomic_signal_fence"] + suffix) and
```

Comment:
We are generating `2 * 2 * 3 = 12` regexes, and a similar number in the other cases. I've seen with other uses of this pattern that it can cause performance problems, so I think we might be better off generating a single regex for each case.

Reply:
However we do this, lets check performance carefully.

Reply:
Changed to just one regex.

This is now the slowest predicate on glfw:

```
        364ms |     2 | 203ms @ 2    | Element::ElementBase.toString/0#dispred#6e016f86@4589bave
```